### ISSUE TITLE
4.1.0 version stopped working on Fedora 33

### ISSUE THREAD

**Issue by tim77** [RadeonOpenCompute/ROCm#1423]
Hello. After today update from 4.0.0 -> 4.1.0 OpenCL not working anymore:

```
❯ rocminfo

ROCk module is loaded
HSA Error:  Incompatible kernel and userspace, Vega 20 [Radeon VII] disabled. Upgrade amdgpu.
=====================    
HSA System Attributes    
=====================    
Runtime Version:         1.1
System Timestamp Freq.:  1000.000000MHz
Sig. Max Wait Duration:  18446744073709551615 (0xFFFFFFFFFFFFFFFF) (timestamp count)
Machine Model:           LARGE                              
System Endianness:       LITTLE                             

==========               
HSA Agents               
==========               
*******                  
Agent 1                  
*******                  
  Name:                    AMD Ryzen 3 3300X 4-Core Processor 
  Uuid:                    CPU-XX                             
  Marketing Name:          AMD Ryzen 3 3300X 4-Core Processor 
  Vendor Name:             CPU                                
  Feature:                 None specified                     
  Profile:                 FULL_PROFILE                       
  Float Round Mode:        NEAR                               
  Max Queue Number:        0(0x0)                             
  Queue Min Size:          0(0x0)                             
  Queue Max Size:          0(0x0)                             
  Queue Type:              MULTI                              
  Node:                    0                                  
  Device Type:             CPU                                
  Cache Info:              
    L1:                      32768(0x8000) KB                   
  Chip ID:                 0(0x0)                             
  Cacheline Size:          64(0x40)                           
  Max Clock Freq. (MHz):   3800                               
  BDFID:                   0                                  
  Internal Node ID:        0                                  
  Compute Unit:            8                                  
  SIMDs per CU:            0                                  
  Shader Engines:          0                                  
  Shader Arrs. per Eng.:   0                                  
  WatchPts on Addr. Ranges:1                                  
  Features:                None
  Pool Info:               
    Pool 1                   
      Segment:                 GLOBAL; FLAGS: KERNARG, FINE GRAINED
      Size:                    8064576(0x7b0e40) KB               
      Allocatable:             TRUE                               
      Alloc Granule:           4KB                                
      Alloc Alignment:         4KB                                
      Accessible by all:       TRUE                               
    Pool 2                   
      Segment:                 GLOBAL; FLAGS: COARSE GRAINED      
      Size:                    8064576(0x7b0e40) KB               
      Allocatable:             TRUE                               
      Alloc Granule:           4KB                                
      Alloc Alignment:         4KB                                
      Accessible by all:       TRUE                               
  ISA Info:                
    N/A                      
*** Done ***             
```

Previous 4.0.0 version work without any issues.

- OS: Fedora 33
- Kernel: 5.11.8
- GPU: Radeon VII
- Mesa: 20.3.4

**Comment by xuhuisheng** [comment]
AMD said it is an already known issue : https://github.com/RadeonOpenCompute/ROCm#bare-metal
Seems upstream kernels cannot compatible with ROCm-4.1.

Could you try install rock-dkms?
Someone said rock-dkms-4.1 based on kernel-5.9.

**Comment by tim77** [comment]
@xuhuisheng thanks, but unfortunately still the same error after installing `rock-dkms-4.1-26.el7.noarch.rpm`:

```
HSA Error:  Incompatible kernel and userspace, Vega 20 [Radeon VII] disabled. Upgrade amdgpu.
```

**Comment by xuhuisheng** [comment]
@tim77 
Do you reboot after installing rock-dkms-4.1?

You can check whether dkms had installed successfully.

```
meicai@meicai-X99:~$ dkms status
amdgpu, 4.1-11, 5.4.0-64-generic, x86_64: installed
amdgpu, 4.1-11, 5.8.0-44-generic, x86_64: installed

```

**Comment by ROCmSupport** [comment]
Hi @tim77 
Thanks for reaching out.
I am unable to reproduce the same issue in my Radeon7 + Ubuntu 18.04.5 when I installed rocm-dkms.

Looks like you have only installed rocm-dev and rock-dkms and missed to install rock-dkms-firmware(firmware component).
Hence I request you to install rock-dkms-firmware also, then reboot and check. It should work well.
Thank you

**Comment by ROCmSupport** [comment]
And also as we mentioned clearly @[https://github.com/RadeonOpenCompute/ROCm#driver-compatibility-issue-in-this-release](url), the ROCm 4.1 run-time and userspace environment are not compatible with ROCm v4.0 and older driver implementations for 7nm-based (Vega 20) hardware (MI50 and MI60).

So request you to clean ROCm rom your machine and install ROCm 4.1 freshly, which solves the problem.
Thank you.

**Comment by preda** [comment]
Could the documentation please specify which minimal Linux kernel version is required for a ROCm 4.1 install *without* dkms?

**Comment by preda** [comment]
With Linux kernel 5.12-rc4, the most recent released ATM, I still get the "upgrade amdgpu" error. It seems not any Linux kernel is fresh enough, i.e. the required amdgpu has not been merged upstream yet?

```
HSA Error:  Incompatible kernel and userspace, Vega 20 [Radeon VII] disabled. Upgrade amdgpu.

Linux x2 5.12.0-051200rc4-generic #202103212230 SMP Sun Mar 21 22:33:27 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux
```

**Comment by darkbasic** [comment]
It works for me, but please notice that I had to manually force upgrade several packages because I was left with a mix of old and new packages.

**Comment by tim77** [comment]
Sorry for some delay here.

> Do you reboot after installing rock-dkms-4.1?

Yes. Installed and rebooted.

> You can check whether dkms had installed successfully.

```
amdgpu, 4.1-26.el7: added
```

> Looks like you have only installed rocm-dev and rock-dkms and missed to install rock-dkms-firmware(firmware component).
Hence I request you to install rock-dkms-firmware also, then reboot and check. It should work well.

I've double-checked and all this components installed: `rocm-dev`, `rock-dkms`, `rock-dkms-firmware`. I've also tried to clean everything and install again as it described in official ROCm docs. Seems like problem come due dkms failed to compile on fresh kernels (`kernel-5.11.9` in my case). Here is a build log (sorry for non-En chars):  [make.log](https://github.com/RadeonOpenCompute/ROCm/files/6198186/make.log)

> It works for me, but please notice that I had to manually force upgrade several packages because I was left with a mix of old and new packages. I'm using an RX 570.

Yep. I did few trick as well to adapt RHEL/CentOS official packages for Fedora. But ROCm v4.0 works flawlessly for me and v4.1 not. And v4.0 works without dkms at all.

Also I've agree with @preda here - would be awesome to have official Linux kernel with amdgpu, instead of dkms. And what is most important here that v4.0 version works without it. I'm crossing my fingers and will hope for the best that AMD will fix this in future.

Maybe my build log could help with fixing dkms build for new kernels. For now i will try to downgrade on previous v4.0. I hope this is possible at least but will be my first time.

**Comment by xuhuisheng** [comment]
@tim77
`added` means not install successfully. You can try `sudo dkms build -m amdgpu -v 4.1-26.el7` and watch if there is any compiling errors.

**Comment by xuhuisheng** [comment]
@tim77
`added` means not install successfully. You can try `sudo dkms build -m amdgpu -v 4.1-26.el7` and watch if there is any compiling errors.

**Comment by ROCmSupport** [comment]
> Yep. I did few tricks as well to adapt RHEL/CentOS official packages for Fedora. But ROCm v4.0 works flawlessly for me and v4.1 not. And v4.0 works without dkms at all.

Hi @tim77, you are referring to ROCm 4.0 works with same 5.11 kernel? **I dont think so.**
Please specify the kernel version you used with ROCm 4.0.

**Comment by ROCmSupport** [comment]
Hi @preda and @tim77 
I would assume that since we're using a 5.9-based kernel, it should work with 5.9 kernel and lower, without dkms.
I tested with 5.8 kernel and it worked well.
As you are using 5.11 and 5.12 kernels, things might not work. Recommend to check with 5.9 and lower versions of kernels.
We clearly mentioned in the docs also that ROCm 4.1 validation is done on 5.4 and 5.6 kernels with dkms, recommend to use the same.
Hope it clarifies.
Thank you.

**Comment by preda** [comment]
@ROCmSupport I have tried ROCm 3.3 and ROCm 4.0 on Linux kernels 5.10, 5.11 and 5.12 and they all work.

OTOH ROCm 4.1 does not work on any of the Linux kernels 5.10, 5.11, 5.12. The error message "Upgrade amdgpu", as I understand now, is also misleading -- as what is required is a *downgrade* of amdgpu to Linux kernel 5.8.

And, given this regression from ROCm 4.0 to ROCm 4.1, and the misleading error message, the result is closing the bug report..

**Comment by darkbasic** [comment]
OpenCL with ROCm 4.1 works fine with my RX 570 on kernel 5.11 (Fedora 33) and no DKMS whatsoever. Here is the proof: https://openbenchmarking.org/result/2103244-AS-ROCM41DAR27

**Comment by acowley** [comment]
As usual, we also have the lower level pieces (OpenCL and HIP) working in NixOS with newer kernels and no dkms. I'm using an RX 580 with kernel 5.10. Binaries are not yet in the main NixOS cache, but are available with the nixos-rocm overlay.

**Comment by darkbasic** [comment]
I've tried tensorflow and it doesn't work: https://github.com/ROCmSoftwarePlatform/tensorflow-upstream/issues/1305
I have no idea if it works with other combinations of ROCm/kernel because it's the first time I try it.

**Comment by tim77** [comment]
@ROCmSupport 
> Hi @tim77, you are referring to ROCm 4.0 works with same 5.11 kernel? I dont think so.
Please specify the kernel version you used with ROCm 4.0.

- kernel 5.11.9 (most recent and up to date kernel).
- `dkms` even not installed.
- OpenCL in ROCm v4.0.1 (`rocm-opencl` package, version `3.6Beta_17_g875c1f8_rocm_rel_4.0_26`) works absolutely fine and stable (except Cycles render in Blender which doesn't work but this is well known issue).

```
  Platform Name:				 AMD Accelerated Parallel Processing
Number of devices:				 1
  Device Type:					 CL_DEVICE_TYPE_GPU
  Vendor ID:					 1002h
  Board name:					 Vega 20 [Radeon VII]
```
Full `clinfo` output: [clinfo.txt](https://github.com/RadeonOpenCompute/ROCm/files/6211080/clinfo.txt).

I've downgraded from `v4.1 -> v4.0.1` and OpenCL works perfectly. Without `dkms`. Screenshot with `hashcat` benchmark for testing purposes:
![Снимок экрана от 2021-03-26 12-23-15](https://user-images.githubusercontent.com/5614476/112618932-5f651800-8e2f-11eb-8888-488f47122fc6.png)

**Comment by ROCmSupport** [comment]
Great to know @tim77, good to know that 4.0 works perfect with 5.11 kernel.
This point made me think little more and allowing me to gather more information.
Let me gather more information and share some update if any.
Thank you.

**Comment by preda** [comment]
@ROCmSupport should maybe the issue be reopened during the investigation.

**Comment by ROCmSupport** [comment]
Looks like we are talking about 2 different issues here.
I request to talk about only one issue for now here and recommend to talk about second issue in a different/new ticket.

ROCm 4.1 + 5.11 kernel + without dkms --> first issue
ROCm 4.1 + 5.11 kernel + with dkms --> second issue
ROCm 4.0 + 5.11 kernel + without dkms --> no issues observed
ROCm 4.0 + 5.11 kernel + with dkms --> no issues observed

Can somebody confirms whether above statements are true.
If true, can you please fill in places of "first issue" and "second issue".
Thank you.

**Comment by ROCmSupport** [comment]
I got some update on "HSA Error:  Incompatible kernel and userspace, Vega 20 [Radeon VII] disabled. Upgrade amdgpu." error:

"Yes, that's the error. Upstream needs the patches to properly report ECC. They can use ROCm 4.1, or build their own custom kernel with the 2 patches included". This issue comes only on VG20 variant cards.
Hence we updated our documentation also with the same.

**Comment by ROCmSupport** [comment]
The patches are in 4.1. No idea when they will make it to upstream. Looks like it will take sometime. So users can't use upstream kernel on Radeon VII with ROCm 4.1. That's the only limitation for now.
Well, we also can't use 4.0 rock-dkms with 4.1 rocm-dev on Radeon VII either too.
So only way for now is: can use rock-dkms + rocm-dev both from 4.1.
And remember one point that its only on Radeon VII. No issues with other asics like Vega10 or MI100 etc.,.
Hope this helps.
Request to file a new issue if any(other than this).

**Comment by preda** [comment]
@ROCmSupport thank you for the clarification!

Please do put out an announcement when the ROCm-4.1 required patches make it into upstream, enabling RadeonVII to be used again with a stock Linux kernel as before.

**Comment by ROCmSupport** [comment]
Sure @preda 
We will update the docs accordingly.
Thank you.

**Comment by grische** [comment]
This seems to be fixed in upstream kernel 5.13: https://github.com/RadeonOpenCompute/ROCm/issues/1478#issuecomment-851189122

I was able to confirm that it works with kernel 5.14. I upgraded the kernel from 5.10.60 to 5.14.21 on the identical system and the error went away.
